modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/model.scala

```diff
@@ -52,26 +58,37 @@ object QState {
     * If the `Current` doesn't have all actions completed or there are no more
     * pending `Execution`s it returns None.
     */
-  def next(st: QState): Option[(Execution[Action], QState)] = for {
-    exe3done <- uncurrentify(st.current)
-    qd = Queue.cons(exe3done)(st.done)
-    (exe3pending, qp) <- Queue.uncons(st.pending)
-    (actions, c) = currentify(exe3pending)
-  } yield (actions, QState(qp, c, qd, st.status))
+  def next(qs: QState): Option[QState] = cleanup(qs) >>= (prime(_))
+
+  // None: current not empty
+  //       pending queue empty
+  def prime(qs: QState): Option[QState] =
+    if (qs.current.isEmpty)
+      Queue.uncons(qs.pending).map(
+        t => {
+         val (exe3, q) = t
+         QState(q, currentify(exe3), qs.done, qs.status)
+        }
+      )
+    else None
 
+  def cleanup(qs: QState): Option[QState] =
+    uncurrentify(qs.current).map(exe3 =>
+      QState(qs.pending, Current.empty, Queue.cons(exe3)(qs.done), qs.status)
+    )
   /**
     * Transform an *unconsed* pending `Execution` into `Current` in addition to
     * returning the unwrapped `Execution`.
     */
   // TODO: Use same structure for `Current` and `Queue.Execution3`?
-  private def currentify(exe3: Queue.Execution3[Action]): (Execution[Action], Current) = {
+  private def currentify(exe3: Queue.Execution3[Action]): (Current) = {
```

Comment:
Right `::` is called `cons` in a list. I just wondered what was the use here. I haven't seen `uncons` though